File: c/binary-search/binary_search.c
#include "binary_search.h"

const int* binary_search(int value, const int* arr, size_t length) {
    if (!arr) return NULL;
    int lo = 0, hi = length;
    while (lo < hi) {
        int mid = (lo + hi) / 2;
        if (arr[mid] == value) {
            return &arr[mid];
        } else if (arr[mid] > value) {
            hi = mid;
        } else {
            lo = mid + 1;
        }
    }
    return NULL;
}

File: zig/binary-search/binary_search.zig
// Take a look at the tests, you might have to change the function arguments

pub fn binarySearch(comptime T: type, target: T, items: []const T) ?usize {
    if (items.len == 0) return null;

    var lo: usize = 0;
    var hi: usize = items.len;
    while (lo < hi) {
        const mid = (lo + hi) / 2;
        if (items[mid] == target) return mid;
        if (items[mid] > target) {
            hi = mid;
        } else {
            lo = mid + 1;
        }
    }
    return null;
}
